src/hotspot/cpu/riscv/stubGenerator_riscv.cpp

```diff
@@ -5322,6 +5322,285 @@ class StubGenerator: public StubCodeGenerator {
+        255u,  26u,  27u,  28u,  29u,  30u,  31u,  32u,  33u,  34u,  35u,  36u,  37u,  38u,  39u,  40u,
+        41u,  42u,  43u,  44u,  45u,  46u,  47u,  48u,  49u,  50u,  51u, 255u, 255u, 255u, 255u, 255u,
+        255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u,
+        255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u,
+        255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u,
+        255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u,
+        255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u,
+        255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u,
+        255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u,
+        255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u, 255u,
+    };
+
+    __ align(CodeEntryAlignment);
+    StubCodeMark mark(this, "StubRoutines", "decodeBlock");
+    address start = __ pc();
+    __ enter();
+
+    Register src    = c_rarg0;
+    Register soff   = c_rarg1;
+    Register send   = c_rarg2;
+    Register dst    = c_rarg3;
+    Register doff   = c_rarg4;
+    Register isURL  = c_rarg5;
+    Register isMIME = c_rarg6;
+
+    Register codec     = c_rarg7;
+    Register dstBackup = x31;
+    Register length    = x28;     // t3, total length of src data in bytes
+
+    Label ProcessData, Exit;
+    Label ProcessScalar, ProcessVector;
+
+    // intrinsic only processes data of length in multiple of 4
+    __ sub(length, send, soff);
+    // it's not guaranteed by java level, so do it explicitly
+    __ andi(length, length, -4);
+    // real src/dst to process data
+    __ add(src, src, soff);
+    __ add(dst, dst, doff);
+    // backup of dst, used to calculate the return value at exit
+    __ mv(dstBackup, dst);
+
+    // load the codec base address
+    __ la(codec, ExternalAddress((address) fromBase64));
+    __ beqz(isURL, ProcessData);
+    __ la(codec, ExternalAddress((address) fromBase64URL));
+    __ BIND(ProcessData);
+
+
+    // scalar version
+    {
+      Label ScalarLoop;
+
+      Register byte0 = soff, byte1 = send, byte2 = doff, byte3 = isURL;
+      Register combined32Bits = x29; // t5
+
+      if (UseRVV) {
+        __ bnez(isMIME, ScalarLoop);
+        __ mv(t0, MaxVectorSize * 4);
+        __ bge(length, t0, ProcessVector);
+        __ BIND(ProcessScalar);
+        __ beqz(length, Exit);
+      }
+
+      __ BIND(ScalarLoop);
+      {
+        // encoded:   [byte0[5:0] : byte1[5:0] : byte2[5:0]] : byte3[5:0]] =>
+        // plain:     [byte0[5:0]+byte1[5:4] : byte1[3:0]+byte2[5:2] : byte2[1:0]+byte3[5:0]]
+
+        // load 4 bytes encoded src data
+        __ lbu(byte0, Address(src, 0));
```

Comment:
Yeh, it could be, I will test it later.

Reply:
there is bit regression with `lw` instead of `lb`.
I think the rational is that lb*4 are not really 4 loads, as they are continuos in address. But with `lw`, it will need other operations to move data from word to 4 bytes respectively.
<google-sheets-html-origin style="caret-color: rgb(0, 0, 0); color: rgb(0, 0, 0); font-style: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: auto; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: auto; word-spacing: 0px; -webkit-text-stroke-width: 0px; text-decoration: none;">
Benchmark | (addSpecial) | (errorIndex) | (lineSize) | (maxNumBytes) | Mode | Cnt | Score -lb | Score -lw | Units | Improvement
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
Base64Decode.testBase64Decode | 0 | 144 | 4 | 1 | avgt |   | 123.03 | 124.96 | ns/op | 1.016
Base64Decode.testBase64Decode | 0 | 144 | 4 | 3 | avgt |   | 146.828 | 145.344 | ns/op | 0.99
Base64Decode.testBase64Decode | 0 | 144 | 4 | 7 | avgt |   | 197.021 | 202.769 | ns/op | 1.029
Base64Decode.testBase64Decode | 0 | 144 | 4 | 32 | avgt |   | 310.964 | 328.367 | ns/op | 1.056
Base64Decode.testBase64Decode | 0 | 144 | 4 | 64 | avgt |   | 432.836 | 464.795 | ns/op | 1.074
Base64Decode.testBase64Decode | 0 | 144 | 4 | 80 | avgt |   | 543.394 | 570.661 | ns/op | 1.05
Base64Decode.testBase64Decode | 0 | 144 | 4 | 96 | avgt |   | 599.538 | 659.938 | ns/op | 1.101
Base64Decode.testBase64Decode | 0 | 144 | 4 | 112 | avgt |   | 714.922 | 793.329 | ns/op | 1.11
Base64Decode.testBase64Decode | 0 | 144 | 4 | 512 | avgt |   | 3054.931 | 3356.059 | ns/op | 1.099
Base64Decode.testBase64Decode | 0 | 144 | 4 | 1000 | avgt |   | 4921.95 | 5413.909 | ns/op | 1.1
Base64Decode.testBase64Decode | 0 | 144 | 4 | 20000 | avgt |   | 78169.374 | 89504.671 | ns/op | 1.145
Base64Decode.testBase64Decode | 0 | 144 | 4 | 50000 | avgt |   | 188385.163 | 218491.724 | ns/op | 1.16

</google-sheets-html-origin>